### ISSUE TITLE
`requires_update==False`: Partially pausing gradient updates of specific nn.Parameter s

### ISSUE THREAD

**Issue by munael** [pytorch/pytorch#51009]
## 🚀 Feature
Introduce a flag `requires_update` or `training` to `nn.Parameter`. Applying `requires_grad && !requires_update` to a parameter in the middle of a network means that gradients propagate as expected through the graph, _but the optimizer is made aware **not** to update the weights of the parameters where `requires_update==False`_.

## Motivation
Support for unconventional training regimes with dynamic freezing of sub-networks, without maintaining a separate optimizer for each "view" of the global network. This workaround is problematic when the views might intersect.

## Pitch
Not sure how this part should differ from ## Feature.

## Alternatives
A generic, but lower level approach that would cover this (maybe?): #38471

## Additional context

N/A

cc @albanD @mruberry @jbschlosser @vincentqb

**Comment by mruberry** [comment]
Thank you for this proposal, @munael. Are there some ways this feature could be added/supported without modifying PyTorch? What would need to be done to do that? I suppose a custom optimizer would need to be implemented, possibly by modifying a current optimizer? It could then look up whether to apply grad updates for each parameter by looking into a dict, for example.

**Comment by munael** [comment]
Impl as a custom optimizer would be fairly straightforward, yes. Though I'd wanted something ready and general when posting this feature request.

It _seems_ like a general enough feature to be a part of the optimizer interface. Anyone knows of an optim example that would somehow meaningfully "break" if told to not update a set of parameters?

**Comment by albanD** [comment]
Hi,

Could you explain why setting `requires_grad=False` (and reseting the .grad field to None) doesn't do what you want here?
This won't change how the gradient propagates through the network in any way.

**Comment by munael** [comment]
@albanD 
> Hi,
> 
> Could you explain why setting `requires_grad=False` (and reseting the .grad field to None) doesn't do what you want here? This won't change how the gradient propagates through the network in any way.

Hi,

...

I've misunderstood (or more like, very smoothly side-stepped) the actual effect and purpose of `requires_grad` by playing with `no_grad()` and mixing unrelated things horribly.

I'm looking into my notes again but should be closing this issue very soon.

Thanks :D
